x-pack/plugins/lens/public/app_plugin/show_underlying_data.ts

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  Query,
+  Filter,
+  DataViewBase,
+  buildCustomFilter,
+  buildEsQuery,
+  FilterStateStore,
+} from '@kbn/es-query';
+import { i18n } from '@kbn/i18n';
+import { RecursiveReadonly } from '@kbn/utility-types';
+import { Capabilities } from 'kibana/public';
+import { TableInspectorAdapter } from '../editor_frame_service/types';
+import { Datasource } from '../types';
+
+export const getShowUnderlyingDataLabel = () =>
+  i18n.translate('xpack.lens.app.openInDiscover', {
+    defaultMessage: 'Open in Discover',
+  });
+
+function joinQueries(queries: Query[][] | undefined) {
+  if (!queries) {
+    return '';
+  }
+  const expression = queries
+    .filter((subQueries) => subQueries.length)
+    .map((subQueries) =>
+      // reduce the amount of round brackets in case of one query
+      subQueries.length > 1
+        ? `( ${subQueries.map(({ query: filterQuery }) => `( ${filterQuery} )`).join(' OR ')} )`
+        : `( ${subQueries[0].query} )`
+    )
+    .join(' AND ');
+  return queries.length > 1 ? `( ${expression} )` : expression;
+}
+
+interface LayerMetaInfo {
+  id: string;
+  columns: string[];
+  filters: {
+    kuery: Query[][] | undefined;
+    lucene: Query[][] | undefined;
+  };
+}
+
+export function getLayerMetaInfo(
+  currentDatasource: Datasource | undefined,
+  datasourceState: unknown,
+  activeData: TableInspectorAdapter | undefined,
+  capabilities: RecursiveReadonly<Capabilities>
+): { meta: LayerMetaInfo | undefined; isVisible: boolean; error: string | undefined } {
+  const isVisible = Boolean(capabilities.navLinks?.discover && capabilities.discover?.show);
+  // If Multiple tables, return
+  // If there are time shifts, return
+  const [datatable, ...otherTables] = Object.values(activeData || {});
+  if (!datatable || !currentDatasource || !datasourceState) {
+    return {
+      meta: undefined,
+      error: i18n.translate('xpack.lens.app.showUnderlyingDataNoData', {
+        defaultMessage: 'Visualization has no data available to show',
+      }),
+      isVisible,
+    };
+  }
+  if (otherTables.length) {
+    return {
+      meta: undefined,
+      error: i18n.translate('xpack.lens.app.showUnderlyingDataMultipleLayers', {
+        defaultMessage: 'Cannot show underlying data for visualizations with multiple layers',
+      }),
+      isVisible,
+    };
+  }
+  const [firstLayerId] = currentDatasource.getLayers(datasourceState);
+  const datasourceAPI = currentDatasource.getPublicAPI({
+    layerId: firstLayerId,
+    state: datasourceState,
+  });
+  // maybe add also datasourceId validation here?
```

Comment:
Let's put it in for now, I think long term we can make it work for other datasources as well.